### ISSUE TITLE
Implement standard hooks in the evaluation loop

### ISSUE THREAD

**Issue by cc-wr** [WolframResearch/WolframLanguageForJupyter#47]
For example:

* `$PreRead`
* `$SyntaxHandler`
* `$Pre`
* `$Post`
* `$PrePrint`

**Comment by cc-wr** [comment]
This is being developed in #46.

**Comment by cc-wr** [comment]
Also see some initial discussion in #45.

**Comment by cc-wr** [comment]
> * `SetOptions[$Output, FormatType -> TraditionalForm]`
> * `$Post=TraditionalForm`
> * `$PrePrint=TraditionalForm`
> * `CurrentValue[$FrontEndSession, {CommonDefaultFormatTypes, "Output"}] = TraditionalForm`

@vsht, #46 should make these work.

(However, the last command only applies to outputs that would have been rasterized by the Front End before `TraditionalForm` was applied. Of course, this can be changed if you or others think such behavior is a bad idea.)

**Comment by kubaPod** [comment]
## $PreRead comment

@cc-wr good catch with the front end origin of the problem. It is not limited to WLfP only, you can reproduce it in Mathematica Kernel.exe as well (if you try to rasterize).

### Questions

1. first question is, how should $PreRead behave, from those examples from Mathematica:


  ![image](https://user-images.githubusercontent.com/15227576/59272190-c3681d80-8c55-11e9-82af-ee478ea020bf.png)

  I'd say it is at least not clear. (not called by ToExpression / called / ?12.?

2. If 'sub parsing' should be affected, and it probably should then user needs to anticipate that $PreRead can't just be a function that will be fed with strings. Otoh is feels pretty annoying to not be able to assume that `$PreRead` will work with strings only in a 'text mode'. 

### Suggestions

- `ToExpression @ frameAssoc.content.code` should respect `$PreRead`, if `ToExpression` uses it it is fine, otherwise it should be added applied to the `code` value.

- **Optionally** there could be a **`$JupyterPreRead`** applied before, which won't interfere with sub parsing. So at the end there would be:
 
 ```Mathematica
totalResult = ToExpression[
  (*$PreRead @*) $JupyterPreRead @ loopState["frameAssoc"]["content"]["code"], 
  InputForm, 
  uninteract
];
```

**Comment by cc-wr** [comment]
I agree with having `WolframLanguageForJupyter` apply `$PreRead` (if it is set) to the input, and letting any sub-`ToExpression`s do what they want.

> * **Optionally** there could be a **`$JupyterPreRead`** applied before, which won't interfere with sub parsing. [...]

My personal opinion on this is disagreement: I want to keep the peculiarities of the `WolframLanguageForJupyter` interface itself to a minimum.

**Comment by kubaPod** [comment]
@cc-wr I don't have a strong opinion on WLFJ specific solutions so we are settled then :)

About applying $PreRead, the question was when ToExpression does it automatically. I don't know exactly but from experiments it seems that it does not do it when then input is an input form string. Which means  that $PreRead should be added manually in all places where content is converted to the code, as code will always be an input form string.

**Comment by cc-wr** [comment]
Currently, `$PreRead` is applied once, in a manner that I think is sufficient, at this location:

https://github.com/WolframResearch/WolframLanguageForJupyter/blob/b63c459fa3dd6356a0f5a22cd3424668b679937d/WolframLanguageForJupyter/Resources/EvaluationUtilities.wl#L300

**Comment by cc-wr** [comment]
#46 should support:

* $PreRead
* $Pre
* $Post
* $PrePrint

I think this is a good stopping point for that pull request, with further additions to be made in succeeding ones.

**Comment by cc-wr** [comment]
A general request for comments on #46 is at #48.

**Comment by cc-wr** [comment]
> * `$PreRead`
> * `$Pre`
> * `$Post`
> * `$PrePrint`

This first set of hooks should now be represented in `master`.
